bigtable/cmd/cbt/cbt_test.go

```diff
@@ -171,3 +178,390 @@ func TestParseColumnsFilter(t *testing.T) {
+		{in: []string{"my-table", "my-file.csv", "batch-size=-5"}, fail: true},
+		{in: []string{"my-table", "my-file.csv", "batch-size=5000000"}, fail: true},
+		{in: []string{"my-table", "my-file.csv", "batch-size=nan"}, fail: true},
+		{in: []string{"my-table", "my-file.csv", "batch-size="}, fail: true},
+		{in: []string{"my-table", "my-file.csv", "workers=0"}, fail: true},
+		{in: []string{"my-table", "my-file.csv", "workers=nan"}, fail: true},
+		{in: []string{"my-table", "my-file.csv", "workers="}, fail: true},
+	}
+	for _, tc := range tests {
+		got, err := parseImporterArgs(context.Background(), tc.in)
+		if !tc.fail && err != nil {
+			t.Errorf("parseImportArgs(%q) unexpectedly failed: %v", tc.in, err)
+			continue
+		}
+		if tc.fail && err == nil {
+			t.Errorf("parseImportArgs(%q) did not fail, out: %q", tc.in, got)
+			continue
+		}
+		if tc.fail {
+			continue
+		}
+		if got.appProfile != tc.out.appProfile ||
+			got.fam != tc.out.fam ||
+			got.sz != tc.out.sz ||
+			got.workers != tc.out.workers {
+			t.Errorf("parseImportArgs(%q) did not fail, out: %q", tc.in, got)
+		}
+	}
+}
+
+func transformToCsvBuffer(data [][]string) ([]byte, error) {
+	if len(data) == 0 {
+		return nil, fmt.Errorf("Data cannot be empty")
+	}
+	var buf bytes.Buffer
+	csvWriter := csv.NewWriter(&buf)
+	if err := csvWriter.WriteAll(data); err != nil {
+		return nil, err
+	}
+	csvWriter.Flush()
+	if err := csvWriter.Error(); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func TestCsvHeaderParser(t *testing.T) {
+	tests := []struct {
+		iData    [][]string
+		iFam     string
+		oFams    []string
+		oCols    []string
+		nextLine []string
+		fail     bool
+	}{
+		// extends "my-family" to col-2
+		{iData: [][]string{{"", "my-family", "", "my-family-2"}, {"", "col-1", "col-2", "col-3"}, {"rk-1", "A", "", ""}},
+			iFam:     "",
+			oFams:    []string{"", "my-family", "my-family", "my-family-2"},
+			oCols:    []string{"", "col-1", "col-2", "col-3"},
+			nextLine: []string{"rk-1", "A", "", ""}},
+		// handles column-faimly=arg-family flag
+		{iData: [][]string{{"", "col-1", "col-2"}, {"rk-1", "A", ""}},
+			iFam:     "arg-family",
+			oFams:    []string{"", "arg-family", "arg-family"},
+			oCols:    []string{"", "col-1", "col-2"},
+			nextLine: []string{"rk-1", "A", ""}},
+
+		// early EOF in headers
+		{iData: [][]string{{"", "my-family", ""}},
+			iFam: "",
+			fail: true},
+	}
+
+	for _, tc := range tests {
+		// create in memory csv like file
+		byteData, err := transformToCsvBuffer(tc.iData)
+		if err != nil {
+			t.Fatal(err)
+		}
+		reader := csv.NewReader(bytes.NewReader(byteData))
+
+		fams, cols, err := parseCsvHeaders(reader, tc.iFam)
+		if !tc.fail && err != nil {
+			t.Errorf("parseCsvHeaders() failed. input:%+v, error:%s", tc, err)
+			continue
+		}
+		if tc.fail && err == nil {
+			t.Errorf("parseImportArgs() did not fail. input:%+v, error:%s", tc, err)
+			continue
+		}
+		if tc.fail {
+			continue
```

Comment:
nit: I think more typically the tests just catch this via an OR or two separate conditionals, collapsing this a bit? I may be missing nuance here

Reply:
This case would allow continuation since the fail is expected, but we don't want to do any comparisons below. It can't be paired with any of the above because they add errors were as this is expected.

I can put the if err != nil {t.Error} inside the tc.fail block,  just wasn't sure if go-style preferred the nested ifs.
```
if tc.fail {
    if err != nil {
	t.Errorf("parseImportArgs() did not fail. input:%+v, error:%s", tc, err)
    }
    continue
}
```

Reply:
this to me seems like we are trying to check if fail end err match. I don't see why they would have different messages. Also one of these has 'parseImportArgs' which I assume is leftover from a copy paste?

Are we in essence checking this
```
if tc.fail != (err != nil)
```

Alternatively, rather than storing tc.fail as a bool, store the err string to do a deeper check: possibly kind of like https://github.com/googleapis/google-cloud-go/blob/master/datastore/datastore_test.go#L1987

Reply:
updated to error matching pattern.